rootfs/etc/nginx/lua/balancer.lua

```diff
@@ -1,86 +1,68 @@
+local sticky = require("balancer.sticky")
 local ewma = require("balancer.ewma")
-local resty_balancer = require("balancer.resty")
 
 -- measured in seconds
 -- for an Nginx worker to pick up the new list of upstream peers
 -- it will take <the delay until controller POSTed the backend object to the Nginx endpoint> + BACKENDS_SYNC_INTERVAL
 local BACKENDS_SYNC_INTERVAL = 1
 
 local DEFAULT_LB_ALG = "round_robin"
+local IMPLEMENTATIONS = {
+  round_robin = round_robin,
+  chash = chash,
+  sticky = sticky,
+  ewma = ewma,
+}
 
 local _M = {}
+local balancers = {}
 
--- TODO(elvinefendi) we can probably avoid storing all backends here. We already store them in their respective
--- load balancer implementations
-local backends, backends_err = lrucache.new(1024)
-if not backends then
-  return error("failed to create the cache for backends: " .. (backends_err or "unknown"))
-end
-
-local function get_current_backend()
-  local backend_name = ngx.var.proxy_upstream_name
-  local backend = backends:get(backend_name)
-
-  if not backend then
-    -- TODO(elvinefendi) maybe force backend sync here?
-    ngx.log(ngx.WARN, "no backend configuration found for " .. tostring(backend_name))
-  end
-
-  return backend
-end
+local function get_implementation(backend)
+  local name = backend["load-balance"] or DEFAULT_LB_ALG
 
-local function get_balancer(backend)
-  if not backend then
-    return nil
+  if backend["sessionAffinityConfig"] and backend["sessionAffinityConfig"]["name"] == "cookie" then
+    name = "sticky"
+  elseif backend["upstream-hash-by"] then
+    name = "chash"
   end
 
-  local lb_alg = backend["load-balance"] or DEFAULT_LB_ALG
-  if resty_balancer.is_applicable(backend) then
-    return resty_balancer
-  elseif lb_alg ~= "ewma" then
-    if lb_alg ~= DEFAULT_LB_ALG then
-      ngx.log(ngx.WARN,
-        string.format("%s is not supported, falling back to %s", backend["load-balance"], DEFAULT_LB_ALG))
-    end
-    return resty_balancer
+  local implementation = IMPLEMENTATIONS[name]
+  if not implementation then
+    ngx.log(ngx.WARN, string.format("%s is not supported, falling back to %s", backend["load-balance"], DEFAULT_LB_ALG))
+    implementation = IMPLEMENTATIONS[DEFAULT_LB_ALG]
   end
 
-  return ewma
-end
-
-local function balance()
-  local backend = get_current_backend()
-  local balancer = get_balancer(backend)
-  if not balancer then
-    return nil, nil
-  end
-
-  local endpoint = balancer.balance(backend)
-  if not endpoint then
-    return nil, nil
-  end
-
-  return endpoint.address, endpoint.port
+  return implementation
 end
 
 local function sync_backend(backend)
-  backends:set(backend.name, backend)
+  local implementation = get_implementation(backend)
+  local balancer = balancers[backend.name]
 
-  local balancer = get_balancer(backend)
   if not balancer then
+    balancers[backend.name] = implementation:new(backend)
+    return
+  end
+
+  if getmetatable(balancer) ~= implementation then
```

Comment:
This is weird .. why are you calling `getmetatable` here?

Reply:
@csfrancis because of the way OOP implemented in Lua. Another option would be to compare `.name` but IMO this is more elegant.

If you notice `new()` function in implementations return a new Lua table and sets its metatable to `self`.

So this logic is to detect the situation where an app starts with let's say EWMA, then in next deploy switches to Round Robin. This condition will detect that and create RR instance.

Reply:
Weird .. that was not immediately obvious to me. More elegant, but I feel like `.name` is more clear. But that's just my opinion, man.

Reply:
Maybe a comment would make this clearer?

Reply:
@csfrancis I wrote this with `.name` initially - another reason why I switched to the current version was that if we use `.name` for this then we have to make that field mandatory for every implementation. Currently `.name` is more for logging purposes and optional.

I'll extract this into a function with a meaningful name :)

Reply:
or maybe just add comment for now